File: src/net/eduard/api/server/minigame/MinigameRoom.java
package net.eduard.api.server.minigame;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.stream.Collectors;

import org.bukkit.entity.Player;

import net.eduard.api.lib.Mine;
import net.eduard.api.lib.storage.StorageAttributes;
import net.eduard.api.lib.storage.Storable;

/**
 * Sala do Minigame
 * 
 * @author Eduard-PC
 *
 */
public class MinigameRoom implements Storable {

	@StorageAttributes(reference = true)
	private Minigame minigame;
	@StorageAttributes(reference = true)
	private MinigameMap map;
	private MinigameMode mode = MinigameMode.NORMAL;
	private int id;
	private int time;
	private boolean enabled;
	private int round;
	private transient MinigameState state = MinigameState.STARTING;
	private transient MinigameMap mapUsed;

	private MinigamePlayer secondWinner;
	private MinigamePlayer thirdWinner;
	private transient List<MinigamePlayer> players = new ArrayList<>();
	private transient List<MinigamePlayer> losers = new ArrayList<>();
	private transient List<MinigameTeam> teams = new ArrayList<>();

	public MinigameRoom() {
	}

	/**
	 * Manda a mensagem para todos os jogadores participando da Sala
	 * 
	 * @param message
	 */
	public void broadcast(String message) {
		for (MinigamePlayer player : players) {
			player.send(minigame.getMessagePrefix()
					+ message.replace("$time", Mine.getTime(time)).replace("$max", "" + getMap().getMaxPlayersAmount())
							.replace("$players", "" + getPlayers().size()));
		}
	}
	public boolean hasMinPlayersAmount() {
		return getPlayers().size() >= getMap().getMinPlayersAmount();
	}
	/**
	 * Verifica se o jogador esta jogando nesta Sala
	 * 
	 * @param player Jogador
	 * @return
	 */
	public boolean isPlaying(Player player) {
		return players.stream().filter(p -> p.getPlayer().equals(player)).findFirst().isPresent();
	}

	public List<Player> getPlayersOnline() {

		return players.stream().filter(p -> p.isOnline()).map(p -> p.getPlayer()).collect(Collectors.toList());
	}

	public List<MinigameTeam> getTeams(MinigamePlayerState state) {
		return teams.stream().filter(t -> t.getPlayers(state).size() > 0).collect(Collectors.toList());
	}

	public MinigameTeam getTeamWinner() {
		return teams.stream().filter(t -> t.getPlayers(MinigamePlayerState.NORMAL).size() > 0).findFirst().get();
	}

	public List<Player> getPlayersOnline(MinigamePlayerState state) {

		return players.stream().filter(p -> p.getState() == state).map(p -> p.getPlayer()).collect(Collectors.toList());
	}

	public List<MinigamePlayer> getPlayers(MinigamePlayerState state) {
		return players.stream().filter(p -> p.getState() == state).collect(Collectors.toList());
	}

	public List<MinigamePlayer> getPlayers() {
		return players;
	}

	public void setPlayers(List<MinigamePlayer> players) {
		this.players = players;
	}

	public void disable() {
		setEnabled(false);
	}

	public void enable() {
		setEnabled(true);
	}

	public boolean checkEnd() {
		return getTime() == getMinigame().getTimeIntoGameOver();
	}

	public boolean checkWinner() {
		return getPlayers(MinigamePlayerState.NORMAL).size() == 1;
	}

	public boolean checkTeamWinner() {
		return teams.stream().filter(team -> team.getPlayers(MinigamePlayerState.NORMAL).size() > 0).count() == 1;
	}

	public MinigamePlayer getWinner() {
		return getPlayers(MinigamePlayerState.NORMAL).get(0);
	}

	public boolean checkForceStart() {
		return getPlayers().size() >= map.getNeededPlayersAmount() && time > getMinigame().getTimeOnForceTimer();
	}

	public void forceGameStart() {
		setTime(getMinigame().getTimeOnForceTimer());
	}

	/**
	 * Aumenta 1 segundo na contagem
	 * 
	 * @return
	 */
	public int advance() {
		return ++time;
	}

	/**
	 * Diminui 1 segundo da contagem
	 * 
	 * @return
	 */
	public int decrease() {
		return --time;
	}

	/**
	 * Coloca o estado desta sala em Jogando (A batalha vai começar)
	 */
	public void startGame() {
		setTime(getMinigame().getTimeOnStartTimer());
		setState(MinigameState.PLAYING);
	}

	/**
	 * Coloca o estado desta sala em Equipando (Pre jogo de muitos minigames)
	 */
	public void startPreGame() {
		setTime(getMinigame().getTimeIntoPlay());
		setState(MinigameState.EQUIPPING);
	}

	/**
	 * Coloca o estado desta sala em Acabando (estado usado em alguns eventos
	 * apenas)
	 */
	public void ending() {
		setTime(getMinigame().getTimeOnRestartTimer());
		setState(MinigameState.ENDING);
	}

	/**
	 * Coloca o estado desta sala em Reiniciando <br>
	 * 'quando fazer eventos use este estado para saber que o evento esta desligado'
	 */
	public void restarting() {
		setState(MinigameState.RESTARTING);
		setTime(getMinigame().getTimeIntoRestart());
	}

	/**
	 * Coloca o estado desta sala em Iniciando
	 */
	public void restart() {
		setState(MinigameState.STARTING);
		setTime(getMinigame().getTimeIntoStart());
	}

	/**
	 * Remove o jogador desta Sala
	 * 
	 * @param player
	 */
	public void leave(MinigamePlayer player) {
		player.getGame().getPlayers().remove(player);
		player.setGame(null);
		for (MinigamePlayer jogador : getPlayers()) {
			jogador.hide(player);
			player.hide(jogador);
		}
	}

	public void leaveAll() {
		Iterator<MinigamePlayer> it = players.iterator();
		while (it.hasNext()) {
			MinigamePlayer player = it.next();
			player.setGame(null);
			it.remove();
		}
	}

	public MinigameRoom(Minigame minigame, MinigameMap map) {
		this.minigame = minigame;
		this.id = minigame.getRooms().size() + 1;
		this.minigame.getRooms().add(this);
		this.map = map;
		this.enabled = true;
		this.time = getMinigame().getTimeIntoStart();
	}

	public int getId() {
		return id;
	}

	public void setId(int id) {
		this.id = id;
	}

	public MinigameState getState() {
		return state;
	}

	public void setState(MinigameState state) {
		this.state = state;
	}

	public MinigameMap getMap() {
		return map;
	}

	public void setMap(MinigameMap map) {
		this.map = map;
	}

	public Minigame getMinigame() {
		return minigame;
	}

	public void setMinigame(Minigame minigame) {
		this.minigame = minigame;
	}

	public int getTime() {
		return time;
	}

	public void setTime(int time) {
		this.time = time;
	}

	public boolean isEnabled() {
		return enabled;
	}

	public void setEnabled(boolean enabled) {
		this.enabled = enabled;
	}

	public List<MinigameTeam> getTeams() {
		return teams;
	}

	public void setTeams(List<MinigameTeam> teams) {
		this.teams = teams;
	}

	public boolean isState(MinigameState state) {
		return this.state == state;
	}

	/**
	 * Remove os jogadores de todos os Times
	 */
	public void emptyTeams() {
		for (MinigameTeam team : teams) {
			team.leaveAll();
		}
	}

	/**
	 * Força a entrada do jogador na Sala
	 * 
	 * @param player Jogador
	 */
	public void join(MinigamePlayer player) {
		player.join(this);
	}

	public MinigameMap getMapUsed() {
		return mapUsed;
	}

	public void setMapUsed(MinigameMap mapUsed) {
		this.mapUsed = mapUsed;
	}

	public MinigameMode getMode() {
		return mode;
	}

	public void setMode(MinigameMode mode) {
		this.mode = mode;
	}

	/**
	 * Verifica se tem Espaço na sala para novos jogadores
	 * 
	 * @return
	 */
	public boolean hasSpace() {
		return getPlayers().size() < getMap().getMaxPlayersAmount();
	}

	public MinigamePlayer getSecondWinner() {
		return secondWinner;
	}

	public void setSecondWinner(MinigamePlayer secondWinner) {
		this.secondWinner = secondWinner;
	}

	public MinigamePlayer getThirdWinner() {
		return thirdWinner;
	}

	public void setThirdWinner(MinigamePlayer thirdWinner) {
		this.thirdWinner = thirdWinner;
	}

	public List<MinigamePlayer> getLosers() {
		return losers;
	}

	public void setLosers(List<MinigamePlayer> losers) {
		this.losers = losers;
	}

	public int getRound() {
		return round;
	}

	public void setRound(int round) {
		this.round = round;
	}

}

File: src/main/java/net/eduard/api/server/minigame/MinigameRoom.kt
package net.eduard.api.server.minigame

import net.eduard.api.lib.database.annotations.ColumnPrimary
import net.eduard.api.lib.modules.Extra

import org.bukkit.entity.Player

import net.eduard.api.lib.storage.annotations.StorageAttributes
import net.eduard.api.lib.storage.annotations.StorageIndex
import net.eduard.api.lib.storage.annotations.StorageReference
import org.bukkit.Bukkit

/**
 * Sala do Minigame
 *
 * @author Eduard-PC
 */
@Suppress("unused")
open class MinigameRoom() {

    @Transient
    lateinit var minigame: Minigame

    @StorageReference
    lateinit var mode: MinigameMode

    @StorageIndex
    @ColumnPrimary
    var id: Int = 0

    @StorageReference
    var map: MinigameMap = MinigameMap()



    open fun start() {
        mapUsed = map.copy()
        mapUsed.minigame = minigame
        mapUsed.worldName = "worlds/${minigame.name}/room/$id"
        if (minigame.worldBased) {
            mapUsed.copyWorld(map)
           // mapUsed.setupChests()
        } else {
            mapUsed.clearWorld()
            mapUsed.paste(mapUsed.feastCenter!!)
        }
        restart()

    }

    open fun stop() {
        mapUsed.unloadWorld()
        restarting()
    }

    open fun reset() {
        if (minigame.worldBased) {
            mapUsed.resetWorld()
           // mapUsed.setupChests()
        } else {
            if (minigame.schematicBasedRestartWorld)
                mapUsed.clearWorld()
            mapUsed.paste(mapUsed.feastCenter!!)
        }
    }

    var isEnabled: Boolean = false
    var ip: String = "127.0.0.1"
    var port: Int = Bukkit.getPort()
    var round: Int = 0
    var state = MinigameState.STARTING
    var time: Int = 0
    var teamSize = 1
    var maxRounds = 3

    @Transient
    var stats = mutableMapOf<MinigamePlayer, MinigamePlayerStats>()

    @Transient
    var mapUsed: MinigameMap = map

    @Transient
    var players = mutableListOf<MinigamePlayer>()

    @Transient
    var losers = mutableListOf<MinigamePlayer>()

    @Transient
    var teams = mutableListOf<MinigameTeam>()

    val playersOnline: List<Player>
        get() = players.filter { it.isOnline }.map { it.player!! }

    val teamWinner: MinigameTeam?
        get() = teams.firstOrNull {
            it.getPlayers(MinigamePlayerState.NORMAL)
                .isNotEmpty()
        }


    val winner: MinigamePlayer
        get() = getPlayers(MinigamePlayerState.NORMAL)[0]


    /**
     * Manda a mensagem para todos os jogadores participando da Sala
     *
     * @param message
     */
    fun broadcast(message: String) {
        for (player in players) {
            player.send(
                minigame.messagePrefix + message
                    .replace("%time", Extra.formatSeconds1(time))
                    .replace("%max", "" + map.maxPlayersAmount)
                    .replace("%players", "" + players.size)
            )
        }
    }

    fun hasMinPlayersAmount(): Boolean {
        return getPlayers(MinigamePlayerState.NORMAL).size >= map.minPlayersAmount
    }

    /**
     * Verifica se o jogador esta jogando nesta Sala
     *
     * @param player Jogador
     * @return
     */
    fun isPlaying(player: Player): Boolean {
        return players.any { minigamePlayer -> minigamePlayer.player == player }

    }

    fun getTeams(state: MinigamePlayerState): List<MinigameTeam> {
        return teams.filter {
            it.getPlayers(state)
                .isNotEmpty()
        }
    }

    fun getPlayersOnline(state: MinigamePlayerState): List<Player> {
        return players.filter { p -> p.state == state && p.isOnline }
            .map { it.player }
    }

    fun getPlayers(state: MinigamePlayerState): List<MinigamePlayer> {
        return players.filter { p -> p.state == state }
    }

    fun checkEnd(): Boolean {
        return time == mode.timeIntoGameOver
    }

    fun checkWinner(): Boolean {
        return getPlayers(MinigamePlayerState.NORMAL).size == 1
    }

    fun checkTeamWinner(): Boolean {
        return teams.filter {
            it.getPlayers(MinigamePlayerState.NORMAL)
                .isNotEmpty()
        }.size == 1
    }

    fun checkForceStart(): Boolean {
        return players.size >= map.neededPlayersAmount &&
                time > mode.timeOnForceTimer
    }

    fun forceGameStart() {
        time = mode.timeOnForceTimer
    }

    /**
     * Aumenta 1 segundo na contagem
     *
     * @return
     */
    fun advance(): Int {
        return ++time
    }

    /**
     * Diminui 1 segundo da contagem
     *
     * @return
     */
    fun decrease(): Int {
        return --time
    }

    /**
     * Coloca o estado desta sala em Jogando (A batalha vai começar)
     */
    fun startGame() {
        time = mode.timeOnStartTimer
        state = MinigameState.PLAYING
    }

    /**
     * Coloca o estado desta sala em Equipando (Pre jogo de muitos minigames)
     */
    fun startPreGame() {
        time = mode.timeIntoPlay
        state = MinigameState.EQUIPPING
    }

    /**
     * Coloca o estado desta sala em Acabando (estado usado em alguns eventos
     * apenas)
     */
    fun ending() {
        time = mode.timeOnRestartTimer
        state = MinigameState.ENDING
    }

    /**
     * Coloca o estado desta sala em Reiniciando <br></br>
     * 'quando fazer eventos use este estado para saber que o evento esta desligado'
     */
    fun restarting() {
        state = MinigameState.RESTARTING
        time = mode.timeIntoRestart
    }

    /**
     * Coloca o estado desta sala em Iniciando
     */
    fun restart() {
        state = MinigameState.STARTING
        time = mode.timeIntoStart
        stats.clear()
    }

    /**
     * Remove o jogador desta Sala
     *
     * @param player
     */
    open fun leave(player: MinigamePlayer) {
        players.remove(player)
        player.resetStats()
        player.game = null
        for (jogador in players) {
            jogador.hide(player)
            player.hide(jogador)
        }
    }

    fun leaveAll() {
        players.forEach { it.game = null }
        players.clear()
    }


    fun isState(state: MinigameState): Boolean {
        return this.state === state
    }

    /**
     * Remove os jogadores de todos os Times
     */
    fun emptyTeams() {
        for (team in teams) {
            team.leaveAll()
        }
        teams.clear()
    }

    /**
     * Força a entrada do jogador na Sala
     *
     * @param player Jogador
     */
    fun join(player: MinigamePlayer) {
        player.join(this)
    }

    /**
     * Verifica se tem Espaço na sala para novos jogadores
     *
     * @return
     */
    fun hasSpace(): Boolean {
        return getPlayers(MinigamePlayerState.NORMAL).size < map.maxPlayersAmount
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as MinigameRoom

        if (id != other.id) return false

        return true
    }

    override fun hashCode(): Int {
        return id
    }


}